### DISCUSSION TITLE
What's the right way to create an encrypted OSD on part of a device?

### DISCUSSION THREAD

**Discussion by dseomn** [rook/rook#7602] [Q&A] [answered]
Hi,

I tried managing the encryption outside of Rook, since my root filesystem is on encrypted LVM anyway, but Rook doesn't seem to support creating volumes on LVs. Then I tried using a raw partition with `encryptedDevice: "true"`, but came across https://github.com/rook/rook/discussions/7349#discussioncomment-425956. If I'm reading https://github.com/rook/rook/pull/4879 correctly, Rook 1.6 will support raw partitions without encryption, but not with encryption. Is that right? Once I upgrade to Rook 1.6, will there be any way to share a single disk with both the system partitions and the OSD, and also encrypt the OSD data? I don't have any real preference between managing the encryption with Rook vs with the system, as long as it's encrypted somehow.

**Comment by leseb** [comment, marked as answer]
It's true that the only scenario where we test consume LVs directly is when deploying the cluster on PVC. So I don't know if it works, AFAIR we skip lvm_member label so what you are saying makes sense.
Raw partition with encryption is not supported.

I'm afraid your scenario is not supported, whether or not the LV is encrypted. Rook at least needs a partition or an entire disk.

**Reply by dseomn** [reply]
Ok, so I think the best solution would be to fix https://tracker.ceph.com/issues/49582 and then change Rook to allow using LVs. Does that make sense to you? Is there a better way forward? (Is this a feature Rook and Ceph want to support at all?)

I looked into that bug a bit and I think I see how to fix it, but I'm waiting for my account on the Ceph Tracker to be approved. Just so I don't forget, I'll leave what I found here for now:

It looks like https://github.com/ceph/ceph/blob/8f57c508df7255bcecbb072d622570c337fdedaa/src/ceph-volume/ceph_volume/util/device.py#L159-L164 is where ceph-volume converts a string argument to info about an LV. When passed `/dev/dm-0`, that's calling a command like `lvs -a -S lv_path=/dev/dm-0` (with some extra args that aren't really relevant here I think). The `lv_path` attribute isn't expecting a `dm-X` device path though, so it doesn't match the device. I ran `lvs -S help` to see the available attributes, and I don't see anything that looks like it could match `/dev/dm-0`, but this works on the command line:

```
root@k8sc0n0:~# stat /dev/dm-0 
  File: /dev/dm-0
  Size: 0               Blocks: 0          IO Block: 4096   block special file
Device: 5h/5d   Inode: 247         Links: 1     Device type: fe,0
Access: (0660/brw-rw----)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2021-04-13 13:58:41.754711923 -0400
Modify: 2021-04-13 13:58:41.754711923 -0400
Change: 2021-04-13 13:58:41.754711923 -0400
 Birth: -
root@k8sc0n0:~# lvs -a -S lv_kernel_major=254,lv_kernel_minor=0
  LV   VG         Attr       LSize  Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  root k8sc0n0-vg -wi-ao---- 64.00g
```

That combination of stat and filtering on the kernel device numbers should be easy enough to do in Python too, I think.

**Reply by leseb** [reply]
Agreed, if this ever works with c-v then we can consider enabling this in Rook.

**Reply by dseomn** [reply]
https://github.com/ceph/ceph/pull/40925 has been stalled waiting for a review for over 2 months. What do you think about changing Rook to pass one of the device paths that already works with Ceph?

**Reply by leseb** [reply]
I asked for some reviews, now it's moving.

**Comment by grafjo** [comment]
@dseomn have you figured out a way to share a single disk with both the system partitions and the OSD, and also encrypt the OSD data or is it still not possible? thanks!

**Reply by dseomn** [reply]
It's been a while since I looked at this (since I gave up on kubernetes entirely for now), but I think this is still blocked on people taking a look at https://github.com/ceph/ceph/pull/40925.

**Reply by grafjo** [reply]
thanks for your fast reply - ok I'll watch the https://github.com/ceph/ceph/pull/40925 so see what is happening there
